Magical mistakes and how to avoid them

audreytoskin opened the pull request:
I'm of course open to feedback if my edits were factually wrong, or you don't like something about the code or prose style, or whathaveyou. I tried to focus:

1. on areas where I was confused while learning Jade or while migrating to Pug.
2. on recurring themes I've seen in invalid Pug issues reported on GitHub.

There isn't currently a fixed order that one is supposed to read the documents, after the "Getting Started" page. For example, on the website, the language reference is sorted alphabetically, but it would probably make sense for most users to read about tags before they read about attributes... This free ordering isn't *necessarily* a bad thing --- it maybe lets users decide what they're interested in learning first. Still, I thought it would help to at least add more cross-references between documentation pages. If you agree, I can do more of that in another pull request.

See messages for individual commits for more details.

joliss committed:
Remove "extends" page (#9)

All of this is covered in more detail on the "Inheritance" page.

audreytoskin committed:
Note common mistakes when working with template inheritance.

Inspired by my issue on the core Pug repo, "Error when there's
unbuffered code at the top of an extending template".

https://github.com/pugjs/pug/issues/2721

In that issue, I guess I'd forgotten that unbuffered code can also
contain Pug markup, because I usually do well enough designing my
template logic around the built-in `if`, `unless`, and `each` keywords.

audreytoskin committed:
Clarify distinction between buffered and unbuffered comments.

The Comments page already described the `//-` tag as an "unbuffered"
comment, so I thought it would help to state the reverse -- that `//`
lines and blocks are "buffered" comments. Inspired by an issue on the
Pug core repo, "Top-level comments in extending templates", where the
issue reporter was apparently confused by the difference between the
two.

https://github.com/pugjs/pug/issues/2632

This will need to be updated if anyone implements the feature suggested
in that thread, where even buffered comments at the top level of an
extending template are simply ignored.

audreytoskin committed:
Expand notes about plain text, inline-level elements, and whitespace.

I rearranged the Plain Text page partly because I disagree with the
statement that piped text is the simplest way of adding plain text.
Whitespace for plain text is pretty tricky for beginners, and I think
piped text in particular seems to come up a lot as a source of
confusion.

So for the Plain Text page, I:

* rearranged the order.
* pointed out the the "weird" whitespace behavior is actually a benefit
for template writers.
* noted that starting a line with literal HTML is also plain text.
* expanded the section about dot-blocks of plain text.
* and wrote a whole section specifically about different ways to control
whitespace in the rendered HTML.

As for the other pages, I mostly just added cross-references where plain
text and whitespace was related. This seems like an important enough
point that we should make sure readers can easily find their way back
here.

audreytoskin committed:
Keep trailing whitespace in Pug and Markdown files.

Trailing whitespace is significant in lines and blocks of plain text and
inline-level tags in Pug. Markdown files that have Pug examples should
therefore also preserve trailing whitespace.

audreytoskin committed:
For tag interpolation add an example containing an attribute.

When I was first getting my head around Pug, I wasn't quite sure where
at first if tag attributes would work or where they should go, in an
interpolated tag.

LionNatsu commented:
Yes, I think good order is needed too. In fact, the order of contents is dependent on `strings.json`, placed at the root->"menu"->[map].

But I won't suggest you to simply change it, because the semantic of map is *unordered*, that is, to modify the order of elements in map is misusing JSON, IMO.

ForbesLindesay commented on src/language/plain-text.md:
```diff
@@ -37,3 +53,78 @@ script.
+
+Depending on where you need the whitespace, you could add an extra space at the beginning of the plain text line (after the pipe for piped text). Or you could add a trailing space at the *end* of the plain text line. Less often, you might need extra space *inside the tag*, but it's the same idea --- extra spaces at the beginning of the text (after the tag), or at the end of the line. This may req ...
+
+**NOTE** the trailing and leading spaces here:
```
Since most editors strip the trailing space by default, let's remove this approach and encourage using one of the other approaches.

audreytoskin replied:
I could definitely try to *discourage* the use of this workaround more, but I don't think it should be removed entirely.

This is indeed sort of a dangerous workaround for users or teams that have not all configured their editors properly. However: 

1. It does work. Users who discover this feature (or side-effect? I wrote about this aspect of Whitespace Control assuming it was a deliberate feature...) may not correctly intuit the rules about leading and trailing whitespace if it's not documented, leading to more annoying "bug" reports.
2. On the other hand, not *all* code editors automatically strip trailing whitespace, and some editor plugins for Pug *prevent* the usual whitespace stripping. In most languages, trailing whitespace is an aesthetic problem only, so it's possible a Pug user might leave trailing whitespace out of negligence and then be surprised by the results.

audreytoskin replied:
Maybe I could split Whitespace Control into subsections: "Recommended Solutions", and "Unrecommended Solutions" or "Dirty Hacks" or something along those lines.

ForbesLindesay commented on src/language/plain-text.md:
````diff
@@ -37,3 +53,78 @@ script.
+| me!
+```
+
+Another trick is to use string [interpolation](interpolation.html): `=` can be used to add a literal space or newline character to the rendered HTML.
````
This is normally called buffered code, not interpolation.  It also seems strictly worse than the `|` approach.

audreytoskin replied:
> buffered code

Ah, right. I'll fix that.

> It also seems strictly worse than the `|` approach.

Funny you should say that. I actually got the idea for this [from the Attributes page](https://pugjs.org/language/attributes.html) that is currently live on the website. The very first example there uses lines of buffered code to separate the inline tags, but does not explain what those `=` lines are for.

audreytoskin replied:
It is admittedly sort of a funny hack. Should we call this a "Discouraged Solution", then, or just remove it?

ForbesLindesay replied:
I think we should probably just remove it.

For the attributes example, what I'd like to do is replace the pug pretty mode with an html formatter based on http://demos.forbeslindesay.co.uk/htmlparser2/ so that we can:

1. deprecate pretty mode, since it's the cause of a lot of people struggling to work out why pug sometimes does one thing and sometimes does something else.
2. change the behaviour of pretty printing to suit our documentation, without worrying about breaking anybody's website.

ForbesLindesay commented on src/language/plain-text.md:
````diff
@@ -37,3 +53,78 @@ script.
+  | line
+```
+
+Especially if your inline tags don't require many attributes, you may find it easiest to use tag interpolation, or literal HTML, within a plain text *block*.
````
Remove the word Especially here

audreytoskin replied:
Done.

ForbesLindesay commented on src/language/tags.md:
```diff
@@ -37,3 +37,7 @@ Tags such as `img`, `meta`, and `link` are automatically self-closing (unless yo
+
+## Rendered Whitespace For Inline Elements
+
+Inline (as opposed to block-level) HTML elements --- that is, elements that only contain ["phrasing content"](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#Phrasing_content) --- behave like [plain text](plain-text.html#whitespace-control) in Pug: By default, whitespace is removed at the beginning and end of the line.
```
This is true of **all** elements, unless the user enables pretty mode, which we actively want to discourage people from using for anything other than generating documentation.

audreytoskin replied:
Oh, I did not realize. I've gotten in the habit of always using "pretty" mode, because it's easier to read the rendered HTML.

For block-level elements, it's not generally a problem for whitespace, though. If a `span` tag and an `a` tag touch in the rendered HTML, the words run together; if two `p` tags touch, they will still be rendered as separate paragraphs in the browser.

audreytoskin replied:
Is there a reason to prefer *not* compiling in "pretty" mode?

ForbesLindesay replied:
Yes, we strongly recommend against pretty mode because of exactly this inconsistency.  It's not clear what is going to be block and what is going to be inline.  A page's style sheet can perfectly legitimately say:

```css
div {
  display: inline;
```

In which case, it may become important to remove the white space between two divs.  If you have pretty mode enabled, this is impossible to do.  It is also very un-safe to swap between pretty mode and non-pretty mode, because it tends to create subtle bugs due to different whitespace behaviour.  We generally recommend using the Elements tab in chrome dev tools (or equivalent in your chosen browser) instead of directly reading the html text that pug outputs.

ForbesLindesay reviewed (changes requested):
This is great, I've just added a few suggestions.

ForbesLindesay commented:
Regarding ordering. I'd like to leave the alphabetical reference, but I'd like to extend the getting started page into a proper tutorial that guides you through what you need to get started with pug.

audreytoskin committed:
Sort Whitespace Control solutions into recommended and not recommended.

`= "\n"` is actually an example of buffered code, not string
interpolation.

audreytoskin committed:
Remove examples of buffered code used as a tag separator.

It was maybe kind of a weird hack anyway.

audreytoskin committed:
Pretty mode is deprecated.

My note about the whitespace behavior for inline-leve tags is actually
also true about block-level tags, except when running in "pretty" mode,
which has been a common source of bugs in users' templates.

audreytoskin commented:
Learning about pretty mode being deprecated, due to the way it changes the output HTML, has somewhat changed my understanding of Pug's whitespace handling. so I'll be sending another commit about this soon...

audreytoskin committed:
Really flesh out the rules for when whitespace is removed or preserved.

Also, if we're actively discouraging people from using pretty mode,
maybe I guess we shouldn't bother with allusions to removing whitespace
"by default." They're better off not enabling pretty mode to put
whitespace back in since it's inconsistent.

ForbesLindesay committed:
grammar

ForbesLindesay committed:
one of the trickiest parts

ForbesLindesay commented:
This is excellent, thanks so much for writing all this up.  I know this will help a lot of people who are getting started with pug.

audreytoskin reviewed (commented):
Note: Using "but" at the beginning of a sentence is perfectly fine, actually. The rule you may often see that says otherwise is based more on the arbitrary whims of grammarians than it is on trying to improve clarity. Moreover, "but" serves essentially the same grammatical purpose as "however" (both are conjunctions that show that the following clause stands in contrast to the text just before it), and people use "however" at the beginning of sentences all the time, even in the most formal writing.

...If you just liked the text here better as a single sentence, though, that's fine too 😜

audreytoskin commented:
Also, I'm glad to be of help. Thanks for all your work on Pug.
